### PR TITLE
Rebase on llvm 16

### PR DESCRIPTION
**Pull request by directhex**
This is a version of https://github.com/dotnet/llvm-project/pull/418 without the merge into dotnet/main, which may be easier for reviewers to parse

### CODE REVIEW

File: llvm/tools/objwriter/objwriter.cpp [hunk 1]
```diff
@@ -228,8 +229,6 @@ MCSection *ObjectWriter::GetSection(const char *SectionName,
     } else {
       Section = ObjFileInfo->getBSSSection();
     } 
-  } else if (strcmp(SectionName, "comment") == 0 && OutContext->getObjectFileType() == MCContext::IsELF) {
-    Section = OutContext->getELFSection(".comment", ELF::SHT_PROGBITS, ELF::SHF_MERGE | ELF::SHF_STRINGS | ELF::SHF_GNU_RETAIN, 1);
```
Comment (filipnavara) [inline llvm/tools/objwriter/objwriter.cpp]: Is there any explanation for why this was removed?

Comment (filipnavara) [reply llvm/tools/objwriter/objwriter.cpp]: Nvm. I see it appears back in the final changes (added in 7a7c0a8a6258e00d14e4bd45739d0c5559abc387).

**Review by filipnavara — APPROVED** [review]
I checked the ObjWriter part only but it looks okay 👍

**Comment by directhex** [comment]
I'm merging this so I can start spitting out some packages into a feed to test against. The PR is big enough that nobody's gonna seriously review it, and we can't afford to sit on our hands if we want stuff which depends on LLVM 16, like source-built emsdk, to go in.

**Pull request merged**